// telegraf-hardened/lib/router.js
"use strict";
/** @format */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const composer_1 = __importDefault(require("./composer"));
/** @deprecated in favor of {@link Composer.dispatch} */
class Router {
    routeFn;
    handlers;
    otherwiseHandler = composer_1.default.passThru();
    constructor(routeFn, handlers = new Map()) {
        this.routeFn = routeFn;
        this.handlers = handlers;
        if (typeof routeFn !== 'function') {
            throw new Error('Missing routing function');
        }
    }
    on(route, ...fns) {
        // siakinnik - removed, typescript guaranties fns.length !== 0
        // if (fns.length === 0) {
        //   throw new TypeError('At least one handler must be provided')
        // };
        for (const fn of fns) {
            if (typeof fn !== 'function' &&
                !(typeof fn === 'object' && 'middleware' in fn)) {
                throw new TypeError(` Router.on handler for route "${route}" must be a function or MiddlewareObj`);
            }
        }
        this.handlers.set(route, composer_1.default.compose(fns));
        return this;
    }
    otherwise(...fns) {
        // siakinnik - removed, typescript guaranties fns.length !== 0
        // if (fns.length === 0) {
        //   throw new TypeError('At least one otherwise handler must be provided')
        // };
        for (const fn of fns) {
            if (typeof fn !== 'function' &&
                !(typeof fn === 'object' && 'middleware' in fn)) {
                throw new TypeError('Telegraf: Router.otherwise handler must be a function or MiddlewareObj');
            }
        }
        this.otherwiseHandler = composer_1.default.compose(fns);
        return this;
    }
    middleware() {
        return composer_1.default.lazy((ctx) => {
            const result = this.routeFn(ctx);
            if (result == null) {
                return this.otherwiseHandler;
            }
            if (result.context) {
                Object.assign(ctx, result.context);
            }
            if (result.state) {
                Object.assign(ctx.state, result.state);
            }
            return this.handlers.get(result.route) ?? this.otherwiseHandler;
        });
    }
}
exports.Router = Router;
