// telegraf-hardened/lib/core/network/client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint @typescript-eslint/restrict-template-expressions: [ "error", { "allowNumber": true, "allowBoolean": true } ] */
const crypto = __importStar(require("crypto"));
const fs = __importStar(require("fs"));
const promises_1 = require("fs/promises");
const path = __importStar(require("path"));
const debug_1 = __importDefault(require("debug"));
const stream_1 = require("stream");
const check_1 = require("../helpers/check");
const compact_1 = require("../helpers/compact");
const multipart_stream_1 = __importDefault(require("./multipart-stream"));
const error_1 = __importDefault(require("./error"));
const url_1 = require("url");
const debug = (0, debug_1.default)('telegraf:client');
const { isStream } = multipart_stream_1.default;
const REQUEST_TIMEOUT = 500000; // ms
async function nativeFetch(url, init) {
    return await globalThis.fetch(url, init);
}
function withTimeout(config, timeout) {
    if (timeout <= 0 || !Number.isFinite(timeout)) {
        return {
            config,
            cleanup: () => undefined,
        };
    }
    const timeoutSignal = AbortSignal.timeout(timeout);
    if (!config.signal) {
        return {
            config: { ...config, signal: timeoutSignal },
            cleanup: () => undefined,
        };
    }
    const abortSignal = AbortSignal;
    if (typeof abortSignal.any === 'function') {
        return {
            config: {
                ...config,
                signal: abortSignal.any([
                    config.signal,
                    timeoutSignal,
                ]),
            },
            cleanup: () => undefined,
        };
    }
    const controller = new AbortController();
    const abort = () => controller.abort();
    const signal = config.signal;
    if (signal.aborted || timeoutSignal.aborted)
        controller.abort();
    else {
        signal.addEventListener('abort', abort, { once: true });
        timeoutSignal.addEventListener('abort', abort, { once: true });
    }
    return {
        config: {
            ...config,
            signal: controller.signal,
        },
        cleanup: () => {
            signal.removeEventListener('abort', abort);
            timeoutSignal.removeEventListener('abort', abort);
        },
    };
}
async function fetchWithTimeout(fetch, url, config, timeout) {
    const request = withTimeout(config, timeout);
    try {
        return await fetch(url, request.config);
    }
    finally {
        request.cleanup();
    }
}
const WEBHOOK_REPLY_METHOD_ALLOWLIST = new Set([
    'answerCallbackQuery',
    'answerInlineQuery',
    'deleteMessage',
    'leaveChat',
    'sendChatAction',
]);
const DEFAULT_EXTENSIONS = {
    audio: 'mp3',
    photo: 'jpg',
    sticker: 'webp',
    video: 'mp4',
    animation: 'mp4',
    video_note: 'mp4',
    voice: 'ogg',
};
const DEFAULT_OPTIONS = {
    apiRoot: 'https://api.telegram.org',
    apiMode: 'bot',
    webhookReply: true,
    testEnv: false,
    fetch: nativeFetch,
    requestTimeout: REQUEST_TIMEOUT,
};
function isInputFile(value) {
    return (!!value &&
        typeof value === 'object' &&
        (((0, check_1.hasProp)(value, 'source') && !!value.source) ||
            ((0, check_1.hasProp)(value, 'url') && !!value.url)));
}
function includesMediaValue(value) {
    if (!value || typeof value !== 'object')
        return false;
    if (Buffer.isBuffer(value) || isStream(value))
        return false;
    if (isInputFile(value))
        return true;
    if (Array.isArray(value))
        return value.some(includesMediaValue);
    return Object.values(value).some(includesMediaValue);
}
function includesMedia(payload) {
    return Object.entries(payload).some(([key, value]) => {
        if (key === 'link_preview_options')
            return false;
        return includesMediaValue(value);
    });
}
function replacer(_, value) {
    if (value == null)
        return undefined;
    return value;
}
function buildJSONConfig(payload) {
    return Promise.resolve({
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(payload, replacer),
    });
}
const FORM_DATA_JSON_FIELDS = [
    'results',
    'reply_markup',
    'mask_position',
    'shipping_options',
    'errors',
];
async function buildFormDataConfig(payload, options) {
    for (const field of FORM_DATA_JSON_FIELDS) {
        if ((0, check_1.hasProp)(payload, field) && typeof payload[field] !== 'string') {
            payload[field] = JSON.stringify(payload[field]);
        }
    }
    const boundary = crypto.randomBytes(32).toString('hex');
    const formData = new multipart_stream_1.default(boundary);
    await Promise.all(Object.keys(payload).map((key) => 
    // @ts-expect-error payload[key] can obviously index payload, but TS doesn't trust us
    attachFormValue(formData, key, payload[key], options)));
    return {
        method: 'POST',
        headers: {
            'content-type': `multipart/form-data; boundary=${boundary}`,
        },
        body: formData,
        duplex: 'half',
    };
}
async function attachFormValue(form, id, value, options) {
    if (value == null) {
        return;
    }
    if (typeof value === 'string' ||
        typeof value === 'boolean' ||
        typeof value === 'number') {
        form.addPart({
            headers: { 'content-disposition': `form-data; name="${id}"` },
            body: `${value}`,
        });
        return;
    }
    if (isInputFile(value)) {
        return await attachFormMedia(form, value, id, options);
    }
    if (Array.isArray(value) || typeof value === 'object') {
        const packedValue = await attachNestedFiles(form, value, options);
        return form.addPart({
            headers: { 'content-disposition': `form-data; name="${id}"` },
            body: JSON.stringify(packedValue),
        });
    }
    return form.addPart({
        headers: { 'content-disposition': `form-data; name="${id}"` },
        body: JSON.stringify(value),
    });
}
async function attachNestedFiles(form, value, options) {
    if (!value || typeof value !== 'object')
        return value;
    if (Buffer.isBuffer(value) || isStream(value))
        return value;
    if (isInputFile(value)) {
        const attachmentId = crypto.randomBytes(16).toString('hex');
        await attachFormMedia(form, value, attachmentId, options);
        return `attach://${attachmentId}`;
    }
    if (Array.isArray(value)) {
        return await Promise.all(value.map((item) => attachNestedFiles(form, item, options)));
    }
    const result = {};
    for (const [key, nestedValue] of Object.entries(value)) {
        result[key] = await attachNestedFiles(form, nestedValue, options);
    }
    return result;
}
function toAttachmentBody(body) {
    if (typeof body === 'string' || Buffer.isBuffer(body) || isStream(body)) {
        return body;
    }
    if (body instanceof ReadableStream) {
        return stream_1.Readable.fromWeb(body);
    }
    throw new TypeError('Unable to read attachment response body');
}
async function attachFormMedia(form, media, id, options) {
    let fileName = media.filename ?? `${id}.${DEFAULT_EXTENSIONS[id] ?? 'dat'}`;
    if ('url' in media && media.url !== undefined) {
        const res = await fetchWithTimeout(options.fetch, media.url, {}, options.requestTimeout);
        if (!res.body)
            throw new TypeError(`Unable to download '${media.url}'`);
        return form.addPart({
            headers: {
                'content-disposition': `form-data; name="${id}"; filename="${fileName}"`,
            },
            body: toAttachmentBody(res.body),
        });
    }
    if ('source' in media && media.source) {
        let mediaSource = media.source;
        if (typeof media.source === 'string') {
            const source = await (0, promises_1.realpath)(media.source);
            if ((await (0, promises_1.stat)(source)).isFile()) {
                fileName = media.filename ?? path.basename(media.source);
                mediaSource = await fs.createReadStream(media.source);
            }
            else {
                throw new TypeError(`Unable to upload '${media.source}', not a file`);
            }
        }
        if (isStream(mediaSource) || Buffer.isBuffer(mediaSource)) {
            form.addPart({
                headers: {
                    'content-disposition': `form-data; name="${id}"; filename="${fileName}"`,
                },
                body: mediaSource,
            });
        }
    }
}
async function answerToWebhook(response, payload, options) {
    if (!includesMedia(payload)) {
        if (!response.headersSent) {
            response.setHeader('content-type', 'application/json');
        }
        response.end(JSON.stringify(payload), 'utf-8');
        return true;
    }
    const { headers, body } = await buildFormDataConfig(payload, options);
    if (!response.headersSent) {
        for (const [key, value] of Object.entries(headers)) {
            response.setHeader(key, value);
        }
    }
    await new Promise((resolve) => {
        response.on('finish', resolve);
        body.pipe(response);
    });
    return true;
}
function setErrorField(error, key, value) {
    try {
        error[key] = value;
        return true;
    }
    catch {
        try {
            Object.defineProperty(error, key, {
                value,
                configurable: true,
                writable: true,
            });
            return true;
        }
        catch {
            return false;
        }
    }
}
function withCause(error, cause) {
    try {
        Object.defineProperty(error, 'cause', {
            value: cause,
            configurable: true,
            writable: true,
        });
    }
    catch {
        // Ignore: this is only a best-effort fallback when redacting native errors.
    }
    return error;
}
function redactToken(error) {
    const redact = (value) => value.replace(/\/(bot|user)(\d+):[^/]+\//, '/$1$2:[REDACTED]/');
    const message = redact(error.message);
    const stack = error.stack ? redact(error.stack) : undefined;
    const redacted = setErrorField(error, 'message', message) &&
        (stack === undefined || setErrorField(error, 'stack', stack));
    if (redacted) {
        throw error;
    }
    const fallback = withCause(new Error(message), error);
    fallback.name = error.name;
    if (stack !== undefined) {
        setErrorField(fallback, 'stack', stack);
    }
    throw fallback;
}
class ApiClient {
    token;
    response;
    options;
    constructor(token, options, response) {
        this.token = token;
        this.response = response;
        const { proxy: proxyOptions, ...apiOptions } = options ?? {};
        this.options = {
            ...DEFAULT_OPTIONS,
            ...(0, compact_1.compactOptions)(apiOptions),
        };
        if (proxyOptions) {
            const { proxy, FetchClient } = proxyOptions;
            if (typeof proxy === 'string' &&
                typeof FetchClient === 'function') {
                const clientInstance = new FetchClient({ proxy });
                this.options.fetch = clientInstance.fetch.bind(clientInstance);
            }
            else {
                throw new Error("Invalid network options: 'proxy' must be a string and 'FetchClient' must be a class.");
            }
        }
    }
    /**
     * If set to `true`, first _eligible_ call will avoid performing a POST request.
     * Note that such a call:
     * 1. cannot report errors or return meaningful values,
     * 2. resolves before bot API has a chance to process it,
     * 3. prematurely confirms the update as processed.
     *
     * https://core.telegram.org/bots/faq#how-can-i-make-requests-in-response-to-updates
     * https://github.com/telegraf/telegraf/pull/1250
     */
    set webhookReply(enable) {
        this.options.webhookReply = enable;
    }
    get webhookReply() {
        return this.options.webhookReply;
    }
    async callApi(method, payload, { signal } = {}) {
        const { token, options, response } = this;
        if (options.webhookReply &&
            response?.writableEnded === false &&
            WEBHOOK_REPLY_METHOD_ALLOWLIST.has(method)) {
            debug('Call via webhook', method, payload);
            // @ts-expect-error using webhookReply is an optimisation that doesn't respond with normal result
            // up to the user to deal with this
            return await answerToWebhook(response, { method, ...payload }, options);
        }
        if (!token) {
            throw new error_1.default({
                error_code: 401,
                description: 'Bot Token is required',
            });
        }
        debug('HTTP call', method, payload);
        const config = includesMedia(payload)
            ? await buildFormDataConfig({ method, ...payload }, options)
            : await buildJSONConfig(payload);
        const apiUrl = new url_1.URL(`./${options.apiMode}${token}${options.testEnv ? '/test' : ''}/${String(method)}`, options.apiRoot);
        config.signal = signal;
        const res = await fetchWithTimeout(options.fetch, apiUrl, config, options.requestTimeout).catch(redactToken);
        if (res.status >= 500) {
            const errorPayload = {
                error_code: res.status,
                description: res.statusText,
            };
            throw new error_1.default(errorPayload, { method, payload });
        }
        const data = (await res.json());
        if (!data.ok) {
            debug('API call failed', data);
            throw new error_1.default(data, { method, payload });
        }
        return data.result;
    }
}
exports.default = ApiClient;
