// telegraf-hardened/lib/core/network/multipart-stream.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const stream = __importStar(require("stream"));
const check_1 = require("../helpers/check");
const CRNL = '\r\n';
class MultipartStream extends stream.Readable {
    boundary;
    parts = [];
    partIterator;
    reading = false;
    constructor(boundary) {
        super();
        this.boundary = boundary;
    }
    addPart(part) {
        this.parts.push(part);
    }
    async *_iterator() {
        for (const [index, part] of this.parts.entries()) {
            yield index === 0
                ? `--${this.boundary}${CRNL}`
                : `${CRNL}--${this.boundary}${CRNL}`;
            for (const [key, header] of Object.entries(part.headers)) {
                yield `${key}:${header}${CRNL}`;
            }
            yield CRNL;
            if (MultipartStream.isStream(part.body)) {
                for await (const chunk of part.body) {
                    yield chunk;
                }
            }
            else {
                yield part.body;
            }
        }
        yield `${CRNL}--${this.boundary}--`;
    }
    _read() {
        this.partIterator ??= this._iterator();
        void this.readNext();
    }
    async readNext() {
        if (this.reading || !this.partIterator)
            return;
        this.reading = true;
        try {
            let shouldContinue = true;
            while (shouldContinue) {
                const { value, done } = await this.partIterator.next();
                if (done) {
                    this.push(null);
                    break;
                }
                shouldContinue = this.push(value);
            }
        }
        catch (error) {
            this.destroy(error);
        }
        finally {
            this.reading = false;
        }
    }
    static isStream(stream) {
        return (typeof stream === 'object' &&
            stream !== null &&
            (0, check_1.hasPropType)(stream, 'pipe', 'function'));
    }
}
exports.default = MultipartStream;
