// telegraf-hardened/lib/core/network/polling.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Polling = void 0;
const debug_1 = __importDefault(require("debug"));
const util_1 = require("util");
const error_1 = require("./error");
const debug = (0, debug_1.default)('telegraf:polling');
const wait = (0, util_1.promisify)(setTimeout);
const DEFAULT_CONFLICT_RETRY_DELAY = 1000;
const DEFAULT_MAX_CONFLICT_RETRY_DELAY = 60000;
function always(x) {
    return () => x;
}
const noop = always(Promise.resolve());
class Polling {
    telegram;
    allowedUpdates;
    options;
    abortController = new AbortController();
    skipOffsetSync = false;
    offset = 0;
    retryCount = 0;
    constructor(telegram, allowedUpdates, options = {}) {
        this.telegram = telegram;
        this.allowedUpdates = allowedUpdates;
        this.options = options;
    }
    async *[Symbol.asyncIterator]() {
        debug('Starting long polling');
        do {
            try {
                const updates = await this.telegram.callApi('getUpdates', {
                    timeout: 50,
                    offset: this.offset,
                    allowed_updates: this.allowedUpdates,
                }, { signal: this.abortController.signal });
                this.retryCount = 0;
                const last = updates[updates.length - 1];
                if (last !== undefined) {
                    this.offset = last.update_id + 1;
                }
                yield updates;
            }
            catch (error) {
                const err = error;
                if (err.name === 'AbortError')
                    return;
                if (err instanceof error_1.TelegramError &&
                    err.code === 409 &&
                    this.options?.retryOnConflict) {
                    const baseDelay = this.options.conflictRetryDelay ??
                        DEFAULT_CONFLICT_RETRY_DELAY;
                    const maxDelay = this.options.maxConflictRetryDelay ??
                        this.options.maxRetryDelay ??
                        DEFAULT_MAX_CONFLICT_RETRY_DELAY;
                    const delay = Math.min(baseDelay * Math.pow(2, this.retryCount++), maxDelay);
                    debug('409 Conflict detected (likely old connection still open). Retrying in %dms (Attempt %d)', delay, this.retryCount);
                    await wait(delay);
                    continue;
                }
                if (err.name === 'FetchError' ||
                    err.message.includes('fetch failed') ||
                    err.code === 'ECONNRESET' ||
                    err.code === 'ETIMEDOUT' ||
                    (err instanceof error_1.TelegramError && err.code === 429) ||
                    (err instanceof error_1.TelegramError && err.code >= 500)) {
                    const retryAfter = err.parameters?.retry_after ?? 5;
                    debug('Failed to fetch updates, retrying after %ds.', retryAfter, err);
                    await wait(retryAfter * 1000);
                    continue;
                }
                if (err instanceof error_1.TelegramError &&
                    // Unauthorized      Conflict
                    (err.code === 401 || err.code === 409)) {
                    this.skipOffsetSync = true;
                    throw err;
                }
                throw err;
            }
        } while (!this.abortController.signal.aborted);
    }
    async syncUpdateOffset() {
        if (this.skipOffsetSync)
            return;
        debug('Syncing update offset...');
        await this.telegram.callApi('getUpdates', {
            offset: this.offset,
            limit: 1,
        });
    }
    async loop(handleUpdate) {
        if (this.abortController.signal.aborted)
            throw new Error('Polling instances must not be reused!');
        try {
            for await (const updates of this)
                await Promise.all(updates.map(handleUpdate));
        }
        finally {
            debug('Long polling stopped');
            // prevent instance reuse
            this.stop();
            await this.syncUpdateOffset().catch(noop);
        }
    }
    stop() {
        this.abortController.abort();
    }
}
exports.Polling = Polling;
